tables/extract.py

```diff
@@ -0,0 +1,131 @@
+        END_TAG (str): The end tag format for identifying tables.
+        workbook (dict[str, pd.DataFrame]): A dictionary containing worksheet names as keys
+            and corresponding pandas DataFrames as values.
+
+    Methods:
+        from_csv(cls, path: Path, worksheet_name: str) -> "TableExtractor":
+            Creates a TableExtractor instance from a CSV file.
+        from_excel(cls, path: Path) -> "TableExtractor":
+            Creates a TableExtractor instance from an Excel file.
+        extract(self, worksheet_name: str, id_tag: str) -> list[Table]:
+            Extracts tables specified by the given ID tag.
+
+    Raises:
+        ValueError: If not all ID tags have matching start and end tags.
+
+    Example usage:
+    >>> extractor = TableExtractor.from_excel(path=...)
+    >>> tables = extractor.extract("Sheet1", "Table1")
+    """
+
+    START_TAG = "{id}S"
+    END_TAG = "{id}E"
+    workbook: dict[str, pd.DataFrame]
+
+    def __init__(self, workbook: dict[str, pd.DataFrame]) -> None:
+        self.workbook = workbook
+
+    @classmethod
+    def from_csv(cls, path: Path, worksheet_name: str) -> "TableExtractor":
+        return cls(workbook={worksheet_name: pd.read_csv(path, header=None, index_col=None)})
+
+    @classmethod
+    def from_excel(cls, path: Path) -> "TableExtractor":
+        return cls(workbook=pd.read_excel(path, index_col=None, header=None, sheet_name=None))
+
+    def extract(self, worksheet_name: str, id_tag: str) -> list[Table]:
+        """Extracts table instances specified by ID.
+
+        Finds IDs in the worksheet that are positioned above the top left and below the bottom right cells of the tables
+        and correctly pairs them together. Using those positions, extracts the table located between them.
+
+        :param worksheet_name: worksheet to extract tables from
+        :param id_tag: ID of table to locate and extract
+        :return: a set of Table objects
+        """
+        worksheet = self.workbook[worksheet_name]
+        end_tags, start_tags = self._get_tags(id_tag, worksheet)
+        paired_tags = self._pair_tags(start_tags, end_tags, file_width=len(worksheet.columns))
+        dfs = self._extract_dfs(worksheet, paired_tags)
+        tables = [Table(df=df, start_tag=start_tag, id_tag=id_tag) for df, (start_tag, _) in zip(dfs, paired_tags)]
+        return tables
+
+    def _get_tags(self, id_tag: str, worksheet: pd.DataFrame) -> tuple[list[Tag], list[Tag]]:
+        start_tag = self.START_TAG.format(id=id_tag)
+        end_tag = self.END_TAG.format(id=id_tag)
+        start_tags = list(zip(*np.where(worksheet == start_tag)))
+        end_tags = list(zip(*np.where(worksheet == end_tag)))
+        start_tags = [Tag(row, col) for row, col in start_tags]
+        end_tags = [Tag(row, col) for row, col in end_tags]
+        if len(start_tags) != len(end_tags):
+            raise ValueError(f"Not all {id_tag} tags have a matching start or end tag.")
```

Comment:
As mentioned in another comment, might be good to have a custom `TableExtractionError` thrown here. Then `process` could throw `TableProcessingError`, and validate could throw its own exception - and all exceptions could sit together in the same `exceptions.py` module